src/World.cpp

```diff
@@ -3289,40 +3339,70 @@ bool cWorld::DoWithPlayerByUUID(const cUUID & a_PlayerUUID, cPlayerListCallback
 
 
 
-cPlayer * cWorld::FindClosestPlayer(Vector3d a_Pos, float a_SightLimit, bool a_CheckLineOfSight)
+
+bool cWorld::DoWithClosestPlayer(Vector3d a_Pos, float a_SightLimit, cPlayerListCallback a_Callback, bool a_CheckLineOfSight, bool a_IgnoreSpectator)
 {
-	double ClosestDistance = a_SightLimit;
-	cPlayer * ClosestPlayer = nullptr;
+	cBoundingBox Box(a_Pos, a_SightLimit * 2);
+	cPlayer * Player = nullptr;
+	double MinimalLength;
 
-	cCSLock Lock(m_CSPlayers);
-	for (cPlayerList::const_iterator itr = m_Players.begin(); itr != m_Players.end(); ++itr)
+	m_ChunkMap->ForEachEntityInBox(Box, [&](cEntity & a_Entity) -> bool
 	{
-		if (!(*itr)->IsTicking())
+		if (a_Entity.IsPlayer())
 		{
-			continue;
-		}
-		Vector3f Pos = (*itr)->GetPosition();
-		double Distance = (Pos - a_Pos).Length();
+			Vector3d Distance = a_Entity.GetPosition() - a_Pos;
+			double Length = Distance.Length();
+			bool Match = false;
+			if (Player == nullptr)  // First match
+			{
+				Match = true;
+			}
+			else
+			{
+				if (MinimalLength > Length)
+				{
+					Match = true;
+				}
+			}
 
-		// If the player is too far, skip them:
-		if (Distance > ClosestDistance)
-		{
-			continue;
-		}
+			if (Match)
+			{
+				cPlayer * PlayerTemp = static_cast<cPlayer *>(&a_Entity);
 
-		// Check LineOfSight, if requested:
-		if (
-			a_CheckLineOfSight &&
-			!cLineBlockTracer::LineOfSightTrace(*this, a_Pos, Pos, cLineBlockTracer::losAirWater)
-		)
-		{
-			continue;
+				// Spectator check
+				if (a_IgnoreSpectator)
+				{
+					if (PlayerTemp->IsGameModeSpectator())
+					{
+						return false;  // Skip
+					}
+				}
+
+				if (a_CheckLineOfSight)
+				{
+					if (!cLineBlockTracer::LineOfSightTrace(*this, a_Pos, a_Entity.GetPosition(), cLineBlockTracer::losAirWater))
+					{
+						return false;  // Skip
+					}
+				}
+
+				MinimalLength = Length;
+				Player = PlayerTemp;  // Save closest player so far
+			}
 		}
 
-		ClosestDistance = Distance;
-		ClosestPlayer = *itr;
+		return false;
+	});
+
+	if (Player == nullptr)
+	{
+		return false;  // Player not found
+	}
+	else
+	{
+		a_Callback(*Player);
+		return true;
```

Comment:
I'm sorry but please put the actual player searching code back to the way it was.  The only change needed to turn this from `FindClosestPlayer` to `DoWithClosestPlayer` was calling the callback at the end of the function.

I'm not a fan of the other changes:
 * `ForEachEntityInBox` has to do more work than looping over players because there are more entities than players
 * the actual player selection code has become more complicated for no apparent reason
 * and most importantly, you've missed the entire point of a `DoWith` function.  It's supposed to call the callback under a lock which isn't the case here because the lock has move into `ForEachEntityInBox`.

Reply:
alright, I agree what you said, I will change based on orignal method.